### ISSUE TITLE
Possible Bug in the Template-Baseline Evaluation?

### ISSUE THREAD

**Issue by kedartatwawadi** [jkkummerfeld/text2sql-data#7]
Hi!

Thanks for providing the datasets and the baseline. 
I think I have encountered a bug in the evaluation of the template baseline, which might be inflating the figures for this baseline:

The[ run_eval()](https://github.com/jkkummerfeld/text2sql-data/blob/master/systems/baseline-template/text2sql-template-baseline.py) function checks if the tags and the template match exactly. If so, it marks the prediction as correct. However, there are some examples, where not all the fields in the template can be directly copied over as 'tags' from the question. Here are some examples: 

```sql
['what', 'are', 'the', 'nonstop', 'flights', 'between', 'SAN', 'JOSE', 'and', 'HOUSTON', 'TEXAS']
['O', 'O', 'O', 'O', 'O', 'O', 'city_name0', 'city_name0', 'O', 'city_name1', 'state_name0']
SELECT DISTINCT FLIGHTalias0.FLIGHT_ID FROM FLIGHT AS FLIGHTalias0 WHERE ( ( FLIGHTalias0.FROM_AIRPORT IN ( SELECT AIRPORT_SERVICEalias0.AIRPORT_CODE FROM AIRPORT_SERVICE AS AIRPORT_SERVICEalias0 WHERE AIRPORT_SERVICEalias0.CITY_CODE IN ( SELECT CITYalias0.CITY_CODE FROM CITY AS CITYalias0 WHERE CITYalias0.CITY_NAME = " city_name0 " ) ) 
AND FLIGHTalias0.TO_AIRPORT IN ( SELECT AIRPORT_SERVICEalias1.AIRPORT_CODE FROM AIRPORT_SERVICE AS AIRPORT_SERVICEalias1 WHERE AIRPORT_SERVICEalias1.CITY_CODE IN ( SELECT CITYalias1.CITY_CODE FROM CITY AS CITYalias1 
WHERE ( CITYalias1.CITY_NAME = " city_name1 " AND CITYalias1.STATE_CODE IN ( SELECT STATEalias0.STATE_CODE FROM STATE AS STATEalias0 WHERE STATEalias0.STATE_NAME = " state_name0 " ) ) ) ) ) AND FLIGHTalias0.STOPS = stops0 ) ;
```
(In this example, stops0 parameter in the SQL query is never filled)

`
```sql 
['los', 'angeles', 'to', 'pittsburgh', 'afternoon', 'tuesday']
['city_name1', 'city_name1', 'O', 'city_name0', 'O', 'day_name0']
SELECT DISTINCT FLIGHTalias0.FLIGHT_ID FROM AIRPORT AS AIRPORTalias0 , AIRPORT AS AIRPORTalias1 , AIRPORT_SERVICE AS AIRPORT_SERVICEalias0 , AIRPORT_SERVICE AS AIRPORT_SERVICEalias1 , CITY AS CITYalias0 , CITY AS CITYalias1 , DAYS AS DAYSalias0 , FLIGHT AS FLIGHTalias0 
WHERE AIRPORTalias0.AIRPORT_CODE = AIRPORT_SERVICEalias0.AIRPORT_CODE AND AIRPORTalias1.AIRPORT_CODE = AIRPORT_SERVICEalias1.AIRPORT_CODE AND CITYalias0.CITY_CODE = AIRPORT_SERVICEalias0.CITY_CODE AND CITYalias0.CITY_NAME = " city_name0 " AND CITYalias1.CITY_CODE = AIRPORT_SERVICEalias1.CITY_CODE AND CITYalias1.CITY_NAME = " city_name1 " AND DAYSalias0.DAY_NAME = " day_name0 " 
AND FLIGHTalias0.DEPARTURE_TIME BETWEEN departure_time0 AND departure_time1 AND FLIGHTalias0.FLIGHT_DAYS = DAYSalias0.DAYS_CODE AND FLIGHTalias0.FROM_AIRPORT = AIRPORTalias1.AIRPORT_CODE AND FLIGHTalias0.TO_AIRPORT = AIRPORTalias0.AIRPORT_CODE ; 
```
(In this example, departure_time0, departure_time1 parameters in the template are never filled)

Is this a bug, or am I missing something? 

`

**Comment by jkkummerfeld** [comment]
A previous issue pointed this out and I thought I fixed it (and rerunning experiments gave essentially the same results), but you are right that there is a bug here that my previous fix didn't cover.

I'm investigating and will follow up.

Thanks for reporting this!

**Comment by jkkummerfeld** [comment]
Just a quick update - I've worked on a fix for this and it looks like it does not substantially alter the results (see the pull request above for details). However, I want to be absolutely sure, so we are working on double checking the new version. If you (@kedartatwawadi) have a chance to look at the new version it would be great to get your thoughts!

**Comment by kedartatwawadi** [comment]
Hi! Thanks for your quick response. I am currently running the updated fix on Advising, Atis datasets, and the evaluation seems to be fixed. Thanks!
(I will get back if I notice anything off)

As for the datasets, I noticed some ambiguous scenarios on the atis dataset. For eg: 


                "text": "city_name1 to city_name0 saturday",
                "question-split": "train",
                "variables": {
                    "city_name0": "BOSTON",
                    "day_number0": "26",
                    "month_number0": "7",
                    "year0": "1991",
                    "city_name1": "PITTSBURGH"
                }
            },
            {
                "text": "list flights from city_name1 to city_name0 on wednesday",
                "question-split": "train",
                "variables": {
                    "city_name0": "DETROIT",
                    "day_number0": "23",
                    "month_number0": "4",
                    "year0": "1991",
                    "city_name1": "PHOENIX"
                }
            },
            {
                "text": "show me flights from city_name1 to city_name0 on sunday",
                "question-split": "train",
                "variables": {
                    "city_name0": "SAN FRANCISCO",
                    "day_number0": "27",
                    "month_number0": "8",
                    "year0": "1991",
                    "city_name1": "PITTSBURGH"
                }
            },
            {
                "text": "show me flights between city_name1 and city_name0 on sunday",
                "question-split": "train",
                "variables": {
                    "city_name0": "LAS VEGAS",
                    "day_number0": "27",
                    "month_number0": "8",
                    "year0": "1991",
                    "city_name1": "NEW YORK"
                }
            },

These examples are quite reasonable, if the model is given an input say "today", but if not then figuring out the date from the day is not possible. It would be great if the dataset improves upon such examples. 

Thanks again for your quick look!

**Comment by jkkummerfeld** [comment]
Yeah, this is a problem in the original ATIS dataset. It was one of a few things we noticed but didn't have the time to fix (on my todo list is to create a file listing these issues for future reference in case someone has time to make improvements).

**Comment by jkkummerfeld** [comment]
I've added the list of issues here:

https://github.com/jkkummerfeld/text2sql-data/blob/master/known-issues.md

**Comment by kedartatwawadi** [comment]
Thanks for creating the list of issues, this is quite helpful!

**Comment by jkkummerfeld** [comment]
Closing this as it is resolved by PR #8
